spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
                 self._add_chunk_to_size(chunk)
                 yield chunk
 
-    def _add_chunk_to_size(self, chunk):
+    def _add_chunk_to_size(self, chunk) -> None:
         self.total_size += len(chunk)
-        if self._maximum_allowed_size is not None and self.total_size > self._maximum_allowed_size:
+        if self.total_size > self._maximum_allowed_size:
             raise ArchiveTooLargeError()
 
-    def read(self, size=-1) -> bytes:
-        content = self._trailing_data
-        self._trailing_data = b''
+    def _num_bytes_left_in_buffer(self) -> int:
+        return len(self._buffer) - self._buffer_position + 1
 
-        if not size or size == -1:
-            for chunk in self._chunks:
-                self._add_chunk_to_size(chunk)
-                content += chunk
-        else:
-            # iterate chunks until we've read at least `size` amount of data
+    def read(self, size=-1) -> bytes | bytearray | memoryview:
+        if self._consumed:
+            return b''
+
+        elif not size or size == -1:
             for chunk in self._chunks:
```

Comment:
The old logic here would _always_ read in at least one more chunk, even if our current buffer had enough data in it. So instead of doing that, we now keep track of how far into our buffer we have read, and only consume more from `self._chunks` when the remaining  number of bytes in the buffer is not enough to satisfy the read `size`